core/variant/array.cpp

```diff
@@ -361,6 +361,89 @@ Array Array::slice(int p_begin, int p_end, int p_step, bool p_deep) const { // l
 	return new_arr;
 }
 
+Array Array::filter(const Callable &p_callable) const {
+	Array new_arr;
+	int true_count = 0;
+	Vector<bool> true_array;
+	true_array.resize(size());
+
+	for (int i = 0; i < size(); i++) {
+		const Variant **argptrs = (const Variant **)alloca(sizeof(Variant *));
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+		p_callable.call(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(Array(), "Error calling method from 'filter': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
+		}
+
+		if (result.operator bool()) {
+			true_array.write[i] = true;
+			true_count++;
+		} else {
+			true_array.write[i] = false;
+		}
+	}
+
+	new_arr.resize(true_count);
+	int j = 0;
+	for (int i = 0; i < size(); i++) {
+		if (true_array[i]) {
+			new_arr[j] = get(i);
+			j++;
+		}
+	}
```

Comment:
I'm not convinced on this second loop. I think it could all be done in a single loop. Instead of storing the `true` or `false` in a vector and then loop again, you can already store the elements on the `new_arr`.

To avoid the multiple allocations on bigger arrays, you could just do `new_arr.resize(size())`, keep an int as the index of the next element, which at the end of the loop will also be a count of filtered elements, so you can resize again to the correct size.